Feature/model autodiscovery

davidorme opened the pull request:
# Description

This PR adds model autodiscovery to the `virtual_rainforest/__init__.py` file. 

Any submodule within `virtual_rainforests/models` will automatically get its schema and BaseModel subclass registered, **as long as** the model `__init__.py` contains the schema registration and imports its BaseModel. The PR also updates the existing `soil` and `abiotic` modules to implement that requirement.

The other advantage here is that the import path for the main models is shorter. That seems like a sensible thing for the core object in each model:

```python
from virtual_rainforest.models.soil import SoilModel
```

rather than:

```python
from virtual_rainforest.models.soil.model import SoilModel
```

Fixes #172 

ETA: Fixes #167

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

davidorme commented:
@dalonsoa @alexdewar

@jacobcook1995 and @vgro and I discussed this recently so I haven't added them as reviewers, but it would be great to get a check for any issues. @TaranRallings is ill.

jacobcook1995 commented:
If possible could you update the documentation in `source/development/defining_new_models.md` to reflect the change?

davidorme commented:
That occurred to me too. I was going to lazily change it when I update the `data` part of the new model definition. But that is lazy.

alexdewar commented on docs/source/development/defining_new_models.md:
````diff
@@ -217,23 +255,3 @@ def schema() -> dict:
 
     return config_schema
 ```
````
Is this code the same for all models, except for the name of the schema file?

If so, you could remove some of the boilerplate and just do something like this in all the `__init__.py` files:

```py
from virtual_rainforest.core.config import default_schema_setup

schema = default_schema_setup()
```

You'll probably need to examine the stack for this.

davidorme replied:
It is the same for all models. That's a good thought.

alexdewar commented on virtual_rainforest/__init__.py:
```diff
@@ -1,13 +1,18 @@
+#   from virtual_rainforest.models.soil.model import SoilModel  # noqa: F401
+# - define a schema member, decorated with @register_schema
+for module_info in pkgutil.iter_modules(vfm.__path__):
+    import_module(f"virtual_rainforest.models.{module_info.name}")
```
...alternatively, you could check here whether the module has an attribute called `schema` and if not, add a default one.

davidorme replied:
One thing that crossed my mind was having a `config.add_module` command which basically wraps this import and schema registration. That could then be used to add custom modules _outside_ of the package tree. If you're at the Python command line, you can always import them, but if you wanted to use them in a config, we'd need a `custom_model` section that could use this too. But, that is probably one for the future.

alexdewar reviewed (approved):
Looks ok to me. I've made a suggestion.

davidorme commented:
@alexdewar @jacobcook1995 

I have updated this PR to implement Alex's suggestion of having a function to setup the schema. I've basically moved all of the reading and validation out of model `__init__.py` and into `config`. I have separated the `register_schema` decorator into `get_schema` function, which handles most of the heavy lifting and error cases, and a `register_schema` function, which just does the registering and wraps `get_schema`. There's a mild update to how we extend the validator.

I've updated the user docs and docstrings and then patched all the existing `__init__.py` files. This also fixes #167.

jacobcook1995 commented on docs/source/development/defining_new_models.md:
```diff
@@ -193,47 +193,63 @@ def cleanup(self) -> None:
+  valid, adds the schema to the schema registry
+  {data}`~virtual_rainforest.core.config.SCHEMA_REGISTRY`.
+1. It also needs to import the main BaseModel subclass. This is not going to be used
+   within the file, so needs `#noqa: 401` to surpress a `flake8` error. This step is
```
suppress

jacobcook1995 commented on docs/source/development/defining_new_models.md:
````diff
@@ -193,47 +193,63 @@ def cleanup(self) -> None:
+
+The `__init__.py` file will then look something like this:
 
 ```python
````
Is it worth adding in a docstring here, as the `__init__.py` now requires a docstring here to document the module correctly (granted this isn't necessarily linked to this pull request)

jacobcook1995 commented on docs/source/development/defining_new_models.md:
```diff
@@ -193,47 +193,63 @@ def cleanup(self) -> None:
+  {data}`~virtual_rainforest.core.config.SCHEMA_REGISTRY`.
+1. It also needs to import the main BaseModel subclass. This is not going to be used
+   within the file, so needs `#noqa: 401` to surpress a `flake8` error. This step is
+   primarily to ensure that the subclass is always loaded from the same location but
```
Not 100% sure I grasp the relevance of "loaded from the same location" isn't it more about ensuring that it's always loaded (so that `pytest` etc are guaranteed to find it)?

davidorme replied:
It was opaque - I've given a bit more detail

jacobcook1995 commented on virtual_rainforest/core/config.py:
```diff
@@ -100,95 +88,122 @@ def log_all_validation_errors(
+        FileNotFoundError: the schema path does not exist
+        JSONDecodeError: the file at the schema path is not valid JSON
+        SchemaError: the file contents are not valid JSON Schema
+        ValueError: the JSON Schema is missing required keys
```
Should these two Value errors be grouped, or are they better left separate?

davidorme replied:
Actually, I think one of them is redundant! Good spot.

jacobcook1995 reviewed (approved):
LGTM, made some minor comments but nothing major
